// spacesvr/main.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectWithoutPropertiesLoose = require('@babel/runtime/helpers/objectWithoutPropertiesLoose');
var reactThreeFiber = require('react-three-fiber');
var THREE = require('three');
var react = require('react');
var drei = require('@react-three/drei');
var PositionalAudioHelper = require('three/examples/jsm/helpers/PositionalAudioHelper');
var RGBELoader = require('three/examples/jsm/loaders/RGBELoader');
var cannon = require('@react-three/cannon');
var ShopifyBuy = require('shopify-buy');
var _extends = require('@babel/runtime/helpers/extends');
var _taggedTemplateLiteralLoose = require('@babel/runtime/helpers/taggedTemplateLiteralLoose');
var core = require('@emotion/core');
var styled = require('@emotion/styled');
var reactDeviceDetect = require('react-device-detect');
var nipplejs = require('nipplejs');
var PointerLockControls = require('three/examples/jsm/controls/PointerLockControls');
var EffectComposer = require('three/examples/jsm/postprocessing/EffectComposer');
var RenderPass = require('three/examples/jsm/postprocessing/RenderPass');
var GlitchPass = require('three/examples/jsm/postprocessing/GlitchPass');
var ShaderPass = require('three/examples/jsm/postprocessing/ShaderPass');
var GammaCorrectionShader = require('three/examples/jsm/shaders/GammaCorrectionShader');
var BloomPass = require('three/examples/jsm/postprocessing/BloomPass');
var FXAAShader = require('three/examples/jsm/shaders/FXAAShader');
var styled$1 = require('@emotion/styled/');
var _asyncToGenerator = require('@babel/runtime/helpers/asyncToGenerator');
var fetch = require('node-fetch');
var DeviceOrientationControls = require('three/examples/jsm/controls/DeviceOrientationControls');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _objectWithoutPropertiesLoose__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutPropertiesLoose);
var ShopifyBuy__default = /*#__PURE__*/_interopDefaultLegacy(ShopifyBuy);
var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _taggedTemplateLiteralLoose__default = /*#__PURE__*/_interopDefaultLegacy(_taggedTemplateLiteralLoose);
var styled__default = /*#__PURE__*/_interopDefaultLegacy(styled);
var nipplejs__default = /*#__PURE__*/_interopDefaultLegacy(nipplejs);
var styled__default$1 = /*#__PURE__*/_interopDefaultLegacy(styled$1);
var _asyncToGenerator__default = /*#__PURE__*/_interopDefaultLegacy(_asyncToGenerator);
var fetch__default = /*#__PURE__*/_interopDefaultLegacy(fetch);

var IMAGE_SRC = "https://d27rt3a60hh1lx.cloudfront.net/images/whiteArrow.png";
var IMAGE_SRC_DARK = "https://d27rt3a60hh1lx.cloudfront.net/images/blackArrow.png";
var Arrow = function Arrow(props) {
  var dark = props.dark,
      restProps = _objectWithoutPropertiesLoose__default['default'](props, ["dark"]);

  var texture = reactThreeFiber.useLoader(THREE.TextureLoader, dark ? IMAGE_SRC_DARK : IMAGE_SRC);
  return /*#__PURE__*/React.createElement("group", restProps, /*#__PURE__*/React.createElement("mesh", {
    scale: [0.004, 0.004, 0.004]
  }, /*#__PURE__*/React.createElement("planeBufferGeometry", {
    attach: "geometry",
    args: [98, 51]
  }), /*#__PURE__*/React.createElement("meshStandardMaterial", {
    map: texture,
    attach: "material",
    alphaTest: 0.5,
    transparent: true,
    normalMap: texture
  })));
};

var environmentStateContext = /*#__PURE__*/react.createContext({});
function useEnvironment() {
  return react.useContext(environmentStateContext);
}
function useEnvironmentState() {
  var _useState = react.useState(true),
      paused = _useState[0],
      setPausedState = _useState[1];

  var _useState2 = react.useState(null),
      overlay = _useState2[0],
      setOverlayState = _useState2[1];

  var container = react.useRef(null);
  var events = react.useRef([]);
  var player = react.useRef({});
  var setPaused = react.useCallback(function (p, o) {
    setPausedState(p);

    if (!p) {
      // if unpausing, set paused overlay to undefined by default
      setOverlayState(o || null);
    }

    if (p && o) {
      // if pausing, only set closed overlay if passed ins
      setOverlayState(o);
    }

    events.current.map(function (ev) {
      if (ev.name === "paused") {
        ev.callback.apply(null, [p, o]);
      }
    });
  }, [events]);

  var setPlayer = function setPlayer(p) {
    player.current = p;
  };

  var addEvent = react.useCallback(function (name, callback) {
    var event = {
      name: name,
      callback: callback
    };
    events.current.push(event);
  }, []);
  var context = {
    paused: paused,
    overlay: overlay,
    player: player.current,
    containerRef: container,
    container: container.current,
    events: events.current,
    setPaused: setPaused,
    setPlayer: setPlayer,
    addEvent: addEvent
  };
  return context;
}
var useControlledProgress = function useControlledProgress() {
  var TIMEOUT = 750;

  var _useProgress = drei.useProgress(),
      progress = _useProgress.progress,
      total = _useProgress.total;

  var startTime = react.useRef(new Date());
  var controlledProgress = react.useRef(0);
  react.useEffect(function () {
    var newTime = new Date();
    var timeElapsed = newTime.getTime() - startTime.current.getTime();
    var diff = Math.min(progress - controlledProgress.current, timeElapsed < TIMEOUT ? 99 : 100);

    if (diff > 0) {
      controlledProgress.current = progress;
    }
  }, [progress]); // wait TIMEOUT (ms) to check if any objects are waiting to be loaded

  var _useState3 = react.useState(0),
      counter = _useState3[0],
      setCounter = _useState3[1];

  var _useState4 = react.useState(false),
      skip = _useState4[0],
      setSkip = _useState4[1];

  react.useEffect(function () {
    if (total > 0) {
      return;
    } else if (counter > 0) {
      setSkip(true);
    } else {
      setTimeout(function () {
        return setCounter(counter + 1);
      }, TIMEOUT);
    }
  }, [counter]);
  return skip ? 100 : Math.floor(controlledProgress.current);
};

var Audio = function Audio(props) {
  var url = props.url,
      _props$position = props.position,
      position = _props$position === void 0 ? new THREE.Vector3(0, 0, 0) : _props$position,
      dCone = props.dCone,
      _props$rollOff = props.rollOff,
      rollOff = _props$rollOff === void 0 ? 1 : _props$rollOff,
      _props$volume = props.volume,
      volume = _props$volume === void 0 ? 7 : _props$volume,
      helper = props.helper;

  var _useEnvironment = useEnvironment(),
      paused = _useEnvironment.paused,
      container = _useEnvironment.container;

  var audioRef = react.useRef();
  var speaker = react.useRef();

  var _useThree = reactThreeFiber.useThree(),
      camera = _useThree.camera,
      scene = _useThree.scene;

  var listener = react.useRef();
  react.useEffect(function () {
    if (container && !audioRef.current) {
      var audio = document.createElement("audio");
      audio.src = url;
      audio.autoplay = false;
      audio.preload = "auto";
      audio.crossOrigin = "anonymous";
      audio.loop = true;
      container == null ? void 0 : container.appendChild(audio);
      audioRef.current = audio;
      return function () {
        audio.pause();

        if (listener.current) {
          camera.remove(listener.current);
        }
      };
    }
  }, [container, audioRef.current]); // audio

  react.useEffect(function () {
    if (!paused && camera && audioRef.current && !speaker.current) {
      listener.current = new THREE.AudioListener();
      camera.add(listener.current);
      speaker.current = new THREE.PositionalAudio(listener.current);
      speaker.current.setMediaElementSource(audioRef.current);
      speaker.current.position.copy(position);

      if (dCone) {
        speaker.current.setDirectionalCone(dCone.x, dCone.y, dCone.z);
      }

      speaker.current.setRolloffFactor(rollOff);
      speaker.current.setVolume(volume);

      if (helper) {
        var _helper = new PositionalAudioHelper.PositionalAudioHelper(speaker.current);

        speaker.current.add(_helper);
      }

      scene.add(speaker.current);
    }
  }, [audioRef.current, camera, speaker.current, listener.current, paused]);
  react.useEffect(function () {
    if (!paused && audioRef.current && audioRef.current.paused) {
      audioRef.current.play();
    }
  }, [paused, audioRef.current]);
  return /*#__PURE__*/React.createElement(React.Fragment, null);
};

var Background = function Background(props) {
  var color = props.color;

  var _useThree = reactThreeFiber.useThree(),
      scene = _useThree.scene;

  var _useState = react.useState(false),
      setup = _useState[0],
      setSetup = _useState[1];

  react.useEffect(function () {
    if (scene && !setup) {
      scene.background = new THREE.Color(color);
      setSetup(true);
    }
  }, [scene, setup]);
  return null;
};

var HDRI = function HDRI(props) {
  var src = props.src;

  var _useThree = reactThreeFiber.useThree(),
      gl = _useThree.gl,
      scene = _useThree.scene; // pmrem generator for hdri loading


  var pmremGenerator = new THREE.PMREMGenerator(gl);
  pmremGenerator.compileEquirectangularShader(); // actual file loader

  var loader = new RGBELoader.RGBELoader();
  loader.setDataType(THREE.UnsignedByteType);
  react.useEffect(function () {
    loader.load(src, function (texture) {
      var envMap = pmremGenerator.fromEquirectangular(texture).texture; // sent envmap onto scene env and background

      scene.environment = envMap;
      scene.background = envMap;
      texture.dispose();
      pmremGenerator.dispose();
    });
  }, [scene, loader, pmremGenerator]);
  return null;
};

var frameWidth = 0.3;
var frameDepth = 0.1;
var borderThickness = 0.2;
var borderDepth = 0.2;
var meshOffset = 0.0005;
var Image = function Image(props) {
  var src = props.src,
      sizeScale = props.sizeScale,
      ratio = props.ratio,
      framed = props.framed,
      doubleSided = props.doubleSided,
      passedMaterial = props.material,
      _props$color = props.color,
      color = _props$color === void 0 ? 0x111111 : _props$color;
  var texture = reactThreeFiber.useLoader(THREE.TextureLoader, src);
  var group = react.useRef();
  var image = react.useRef(); // sizing

  var normalizedRatio = new THREE.Vector2(ratio[0], ratio[1]).normalize();
  var width = normalizedRatio.x * sizeScale;
  var height = normalizedRatio.y * sizeScale;
  var material = react.useMemo(function () {
    return passedMaterial || new THREE.MeshStandardMaterial({
      color: color,
      roughness: 0.8,
      metalness: 0.05
    });
  }, []);
  return /*#__PURE__*/React.createElement("group", props, /*#__PURE__*/React.createElement("group", {
    ref: group
  }, /*#__PURE__*/React.createElement("mesh", {
    castShadow: true,
    ref: image
  }, /*#__PURE__*/React.createElement("planeBufferGeometry", {
    attach: "geometry",
    args: [width, height]
  }), /*#__PURE__*/React.createElement("meshStandardMaterial", {
    attach: "material",
    map: texture,
    side: doubleSided ? THREE.DoubleSide : undefined
  })), framed && /*#__PURE__*/React.createElement(React.Fragment, null, !doubleSided && /*#__PURE__*/React.createElement("mesh", {
    "position-z": [-0.1 - meshOffset],
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [width + frameWidth, height + frameWidth, frameDepth]
  })), /*#__PURE__*/React.createElement("mesh", {
    "position-y": height / 2 + frameWidth / 2 - borderThickness / 2,
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [width + frameWidth, borderThickness, borderDepth]
  })), /*#__PURE__*/React.createElement("mesh", {
    "position-y": -height / 2 - frameWidth / 2 + borderThickness / 2,
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [width + frameWidth, borderThickness, borderDepth]
  })), /*#__PURE__*/React.createElement("mesh", {
    "position-x": -width / 2 - frameWidth / 2 + borderThickness / 2,
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [borderThickness, height + frameWidth, borderDepth]
  })), /*#__PURE__*/React.createElement("mesh", {
    "position-x": width / 2 + frameWidth / 2 - borderThickness / 2,
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [borderThickness, height + frameWidth, borderDepth]
  })))));
};

var InfinitePlane = function InfinitePlane(props) {
  var height = props.height,
      _props$size = props.size,
      size = _props$size === void 0 ? [100, 100] : _props$size,
      visible = props.visible;

  var _usePlane = cannon.usePlane(function () {
    return {
      rotation: [-Math.PI / 2, 0, 0],
      position: [0, height, 0],
      args: size,
      type: "Static"
    };
  }),
      ref = _usePlane[0];

  return /*#__PURE__*/React.createElement("mesh", {
    ref: ref
  }, visible && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("planeBufferGeometry", {
    attach: "geometry",
    args: size
  }), /*#__PURE__*/React.createElement("meshPhongMaterial", {
    attach: "material",
    color: "#660000"
  })));
};

var SPEED = 0.2;
var FILE_URL = "https://d27rt3a60hh1lx.cloudfront.net/models/SpacesSphere1/SpacesSphere1.glb";
var Logo = function Logo(props) {
  var rotating = props.rotating,
      floating = props.floating,
      restProps = _objectWithoutPropertiesLoose__default['default'](props, ["rotating", "floating"]);

  var group = react.useRef();
  var sphereGroup = react.useRef();

  var _ref = drei.useGLTF(FILE_URL),
      nodes = _ref.nodes,
      materials = _ref.materials;

  reactThreeFiber.useFrame(function (_ref2) {
    var clock = _ref2.clock;

    if (group.current && rotating) {
      group.current.rotation.y = clock.getElapsedTime() * SPEED;
    }

    if (group.current && floating) {
      group.current.position.y = 0.1 * 2 * Math.sin(clock.getElapsedTime() * SPEED * 2);
    }
  });
  return /*#__PURE__*/React.createElement("group", restProps, /*#__PURE__*/React.createElement("group", {
    ref: group
  }, /*#__PURE__*/React.createElement(react.Suspense, {
    fallback: null
  }, /*#__PURE__*/React.createElement("group", {
    ref: sphereGroup,
    scale: [100, 100, 100],
    dispose: null
  }, /*#__PURE__*/React.createElement("group", {
    position: [0, 0, 0]
  }, /*#__PURE__*/React.createElement("mesh", {
    material: materials.Sphere,
    geometry: nodes.Sphere.geometry,
    castShadow: true
  }))))));
};
drei.useGLTF.preload(FILE_URL);

var useTrimeshCollision = function useTrimeshCollision(geometry) {
  var vertices = geometry.attributes.position.array; // @ts-ignore

  var indices = geometry.index.array; // @ts-ignore

  var _useTrimesh = cannon.useTrimesh(function () {
    return {
      type: "Static",
      args: [vertices, indices]
    };
  }),
      hitbox = _useTrimesh[0];

  return hitbox;
};
var useConvexCollision = function useConvexCollision(geometry) {
  var geo = react.useMemo(function () {
    return new THREE.Geometry().fromBufferGeometry(geometry);
  }, [geometry]);

  var _useConvexPolyhedron = cannon.useConvexPolyhedron(function () {
    return {
      type: "Static",
      args: geo
    };
  }),
      hitbox = _useConvexPolyhedron[0];

  return hitbox;
};

var useShopify = function useShopify(props) {
  var domain = props.domain,
      token = props.token;
  var client = ShopifyBuy__default['default'].buildClient({
    domain: domain,
    storefrontAccessToken: token
  });

  var _useState = react.useState(),
      products = _useState[0],
      setProducts = _useState[1];

  var _useState2 = react.useState(),
      checkout = _useState2[0],
      setCheckout = _useState2[1];

  var _useState3 = react.useState("false"),
      checkoutOpen = _useState3[0],
      setCheckoutOpen = _useState3[1];

  react.useEffect(function () {
    if (!products && !checkout) {
      client.product.fetchAll() // @ts-ignore
      .then(function (shopifyProducts) {
        return setProducts(shopifyProducts);
      });
      client.checkout.create() // @ts-ignore
      .then(function (shopifyCheckout) {
        return setCheckout(shopifyCheckout);
      });
    }
  }, [products, checkout]);
  return {
    client: client,
    products: products,
    checkout: checkout,
    setCheckout: setCheckout,
    checkoutOpen: checkoutOpen,
    setCheckoutOpen: setCheckoutOpen
  };
};

/**
 *
 * Interactible adds on click and hover methods to any group of Object3D's
 *
 * @param props
 * @constructor
 */
var Interactable = function Interactable(props) {
  var children = props.children,
      onClick = props.onClick,
      onHover = props.onHover,
      onUnHover = props.onUnHover;

  var _useEnvironment = useEnvironment(),
      player = _useEnvironment.player;

  var raycaster = player.raycaster;
  var group = react.useRef();

  var _useState = react.useState(false),
      hovered = _useState[0],
      setHovered = _useState[1];

  reactThreeFiber.useFrame(function () {
    if (group.current && raycaster) {
      var intersections = raycaster.intersectObject(group.current, true);

      if (intersections && intersections.length > 0) {
        if (!hovered) {
          setHovered(true);

          if (onHover) {
            onHover();
          }
        }
      } else if (hovered) {
        setHovered(false);

        if (onUnHover) {
          onUnHover();
        }
      }
    }
  });
  react.useEffect(function () {
    var checkClick = function checkClick() {
      if (hovered && onClick) {
        onClick();
      }
    };

    document.addEventListener("click", checkClick);
    return function () {
      document.removeEventListener("click", checkClick);
    };
  }, [hovered, onClick]);
  return /*#__PURE__*/React.createElement("group", {
    ref: group
  }, children);
};

var Floating = function Floating(props) {
  var children = props.children,
      _props$height = props.height,
      height = _props$height === void 0 ? 0.2 : _props$height,
      _props$speed = props.speed,
      speed = _props$speed === void 0 ? 1 : _props$speed;
  var group = react.useRef();
  var seed = react.useRef(Math.random());
  reactThreeFiber.useFrame(function (_ref) {
    var clock = _ref.clock;

    if (group.current) {
      group.current.position.y = height * Math.sin(clock.getElapsedTime() * speed * 0.4 + seed.current * 10000);
    }
  });
  return /*#__PURE__*/React.createElement("group", {
    ref: group
  }, children);
};

function _createForOfIteratorHelperLoose(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; return function () { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } it = o[Symbol.iterator](); return it.next.bind(it); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
var Shop = function Shop(props) {
  var domain = props.domain,
      token = props.token,
      localProducts = props.localProducts,
      _props$itemSize = props.itemSize,
      itemSize = _props$itemSize === void 0 ? 4 : _props$itemSize,
      _props$itemRatio = props.itemRatio,
      itemRatio = _props$itemRatio === void 0 ? [1, 1] : _props$itemRatio,
      _props$spaceBetween = props.spaceBetween,
      spaceBetween = _props$spaceBetween === void 0 ? 9 : _props$spaceBetween,
      _props$position = props.position,
      position = _props$position === void 0 ? [-9, 4, 0] : _props$position,
      _props$rotation = props.rotation,
      rotation = _props$rotation === void 0 ? [0, Math.PI / 9, 0] : _props$rotation;
  var productImages = [];
  var positionOffset = 0;
  var offset = 0;

  if (localProducts) {
    var _loop = function _loop() {
      var product = _step.value;
      var productImage = /*#__PURE__*/React.createElement(Interactable, {
        onClick: function onClick() {
          return window.location.href = product[1];
        },
        key: product[0]
      }, /*#__PURE__*/React.createElement(Image, {
        src: product[0],
        ratio: itemRatio,
        sizeScale: itemSize,
        position: [Math.cos(offset) * 19 + spaceBetween, 2, Math.sin(offset) * 19],
        rotation: [0, Math.PI / 2 - offset + Math.PI, 0],
        framed: true
      }));
      productImages.push(productImage);
      offset += 2 * Math.PI / localProducts.length;
      positionOffset++;
    };

    for (var _iterator = _createForOfIteratorHelperLoose(localProducts), _step; !(_step = _iterator()).done;) {
      _loop();
    }

    return /*#__PURE__*/React.createElement("group", {
      position: position,
      rotation: rotation
    }, productImages);
  }

  var shopifyState;

  if (domain && token) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    shopifyState = useShopify({
      domain: domain,
      token: token
    });
  } // @ts-ignore


  var _shopifyState = shopifyState,
      products = _shopifyState.products;

  if (products) {
    // @ts-ignore
    products.forEach(function (product) {
      if (product.images[0]) {
        var src = product.images[0].src;
        var productImage = /*#__PURE__*/React.createElement(Image, {
          src: src,
          ratio: itemRatio,
          sizeScale: itemSize,
          position: [spaceBetween * positionOffset, 2, 0],
          rotation: [0, Math.PI, 0],
          framed: true
        });
        productImages.push(productImage);
        positionOffset++;
      }
    });
  }

  return /*#__PURE__*/React.createElement("group", {
    position: position,
    rotation: rotation
  }, productImages);
};

var FONT_FILE = "fonts/Couture_Bold.json";
var Text = function Text(props) {
  var text = props.text,
      _props$vAlign = props.vAlign,
      vAlign = _props$vAlign === void 0 ? "center" : _props$vAlign,
      _props$hAlign = props.hAlign,
      hAlign = _props$hAlign === void 0 ? "center" : _props$hAlign,
      _props$size = props.size,
      size = _props$size === void 0 ? 1 : _props$size,
      _props$bevel = props.bevel,
      bevel = _props$bevel === void 0 ? false : _props$bevel,
      _props$color = props.color,
      color = _props$color === void 0 ? "#000000" : _props$color,
      fontFile = props.font,
      material = props.material,
      restProps = _objectWithoutPropertiesLoose__default['default'](props, ["text", "vAlign", "hAlign", "size", "bevel", "color", "font", "material"]);

  var font = reactThreeFiber.useLoader(THREE.FontLoader, fontFile || FONT_FILE);
  var config = react.useMemo(function () {
    return {
      font: font,
      size: size,
      height: 0.75 * size,
      curveSegments: 32,
      bevelEnabled: bevel,
      bevelThickness: 0.15 * size,
      bevelSize: 0.0625 * size,
      bevelOffset: 0,
      bevelSegments: 8
    };
  }, [font]);
  var mesh = reactThreeFiber.useUpdate(function (self) {
    var size = new THREE.Vector3();
    self.geometry.computeBoundingBox();
    self.geometry.boundingBox.getSize(size);
    self.position.x = hAlign === "center" ? -size.x / 2 : hAlign === "right" ? 0 : -size.x;
    self.position.y = vAlign === "center" ? -size.y / 2 : vAlign === "top" ? 0 : -size.y;
  }, [text]);
  return /*#__PURE__*/React.createElement("group", _extends__default['default']({}, restProps, {
    scale: [0.1 * size, 0.1 * size, 0.1]
  }), /*#__PURE__*/React.createElement("mesh", {
    ref: mesh,
    material: material
  }, /*#__PURE__*/React.createElement("textGeometry", {
    attach: "geometry",
    args: [text, config]
  }), !material && /*#__PURE__*/React.createElement("meshPhongMaterial", {
    attach: "material",
    color: color,
    reflectivity: 30
  })));
};

var frameWidth$1 = 0.3;
var frameDepth$1 = 0.1;
var borderThickness$1 = 0.2;
var borderDepth$1 = 0.2;
var meshOffset$1 = 0.0005;
var Video = function Video(props) {
  var src = props.src,
      sizeScale = props.sizeScale,
      ratio = props.ratio,
      framed = props.framed,
      position = props.position,
      rotation = props.rotation,
      muted = props.muted,
      doubleSided = props.doubleSided;

  var _useThree = reactThreeFiber.useThree(),
      camera = _useThree.camera,
      scene = _useThree.scene;

  var _useEnvironment = useEnvironment(),
      container = _useEnvironment.container,
      paused = _useEnvironment.paused;

  var group = react.useRef(); // video state

  var videoRef = react.useRef();
  var textureRef = react.useRef();

  var _useState = react.useState(false),
      texReady = _useState[0],
      setTexReady = _useState[1]; // audio refs


  var listener = react.useRef();
  var speaker = react.useRef();
  reactThreeFiber.useFrame(function () {
    var _videoRef$current;

    if (!texReady && (videoRef == null ? void 0 : videoRef.current) && (videoRef == null ? void 0 : (_videoRef$current = videoRef.current) == null ? void 0 : _videoRef$current.currentTime) > 0) {
      setTexReady(true);
    }
  });
  var material = react.useMemo(function () {
    return new THREE.MeshStandardMaterial({
      color: 0x111111,
      roughness: 0.8,
      metalness: 0.05
    });
  }, []); // sizing

  var normalizedRatio = new THREE.Vector2(ratio[0], ratio[1]).normalize();
  var width = normalizedRatio.x * sizeScale;
  var height = normalizedRatio.y * sizeScale; // video textures use effect

  react.useEffect(function () {
    if (container && !videoRef.current) {
      // build video dom element
      var video = document.createElement("video");
      var source = document.createElement("source");
      source.src = src;
      video.loop = true; //@ts-ignore

      video.playsInline = true;
      video.crossOrigin = "anonymous";
      video.preload = "auto";
      video.autoplay = false;
      video.muted = muted || false;
      video.style.position = "absolute";
      video.style.opacity = "0";
      video.style.pointerEvents = "none";
      video.style.visibility = "hidden"; // add to parent container

      container.appendChild(video);
      video.appendChild(source); // create video texture

      var texture = new THREE.VideoTexture(video);
      texture.minFilter = THREE.LinearFilter;
      texture.magFilter = THREE.LinearFilter;
      texture.format = THREE.RGBFormat;
      texture.matrixAutoUpdate = false;
      texture.wrapS = THREE.ClampToEdgeWrapping;
      texture.wrapT = THREE.ClampToEdgeWrapping;
      videoRef.current = video;
      textureRef.current = texture;
      return function () {
        video.pause();
        video.remove();
        videoRef.current = undefined;
        texture.dispose();
        textureRef.current = undefined;

        if (speaker.current) {
          speaker.current.disconnect();
          speaker.current = undefined;
        }

        if (listener.current) {
          camera.remove(listener.current);
        }

        setTexReady(false);
      };
    }
  }, [container, videoRef == null ? void 0 : videoRef.current]); // audio useeffect

  react.useEffect(function () {
    if (!muted && !paused && camera && videoRef.current && !speaker.current) {
      listener.current = new THREE.AudioListener();
      camera.add(listener.current);
      speaker.current = new THREE.PositionalAudio(listener.current);
      speaker.current.setMediaElementSource(videoRef.current);
      speaker.current.setRefDistance(0.75);
      speaker.current.setRolloffFactor(1);
      speaker.current.setVolume(1);
      speaker.current.setDirectionalCone(180, 230, 0.1);

      if (position) {
        // @ts-ignore due to bad types, it's an array though
        speaker.current.position.set(position[0], position[1], position[2]);
      }

      if (rotation) {
        // @ts-ignore due to bad types, it's an array though
        speaker.current.rotation.set(rotation[0], rotation[1], rotation[2]);
      }

      scene.add(speaker.current);
    }
  }, [videoRef.current, camera, speaker.current, listener.current, paused]); // play video if player closes menu

  react.useEffect(function () {
    if (!paused && videoRef.current && videoRef.current.paused) {
      videoRef.current.play();
    }
  }, [paused]);
  return /*#__PURE__*/React.createElement("group", props, /*#__PURE__*/React.createElement("group", {
    ref: group
  }, texReady && /*#__PURE__*/React.createElement("mesh", null, /*#__PURE__*/React.createElement("planeBufferGeometry", {
    attach: "geometry",
    args: [width, height]
  }), /*#__PURE__*/React.createElement("meshStandardMaterial", {
    attach: "material",
    map: textureRef.current,
    side: doubleSided ? THREE.DoubleSide : undefined
  })), framed && /*#__PURE__*/React.createElement(React.Fragment, null, !doubleSided && /*#__PURE__*/React.createElement("mesh", {
    "position-z": [-0.1 - meshOffset$1],
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [width + frameWidth$1, height + frameWidth$1, frameDepth$1]
  })), /*#__PURE__*/React.createElement("mesh", {
    "position-y": height / 2 + frameWidth$1 / 2 - borderThickness$1 / 2,
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [width + frameWidth$1, borderThickness$1, borderDepth$1]
  })), /*#__PURE__*/React.createElement("mesh", {
    "position-y": -height / 2 - frameWidth$1 / 2 + borderThickness$1 / 2,
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [width + frameWidth$1, borderThickness$1, borderDepth$1]
  })), /*#__PURE__*/React.createElement("mesh", {
    "position-x": -width / 2 - frameWidth$1 / 2 + borderThickness$1 / 2,
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [borderThickness$1, height + frameWidth$1, borderDepth$1]
  })), /*#__PURE__*/React.createElement("mesh", {
    "position-x": width / 2 + frameWidth$1 / 2 - borderThickness$1 / 2,
    material: material
  }, /*#__PURE__*/React.createElement("boxBufferGeometry", {
    attach: "geometry",
    args: [borderThickness$1, height + frameWidth$1, borderDepth$1]
  })))));
};

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: absolute;\n  width: 90%;\n  height: 90%;\n  max-width: 500px;\n  max-height: 500px;\n  background-image: url(", ");\n  background-size: contain;\n  background-position: center;\n  background-repeat: no-repeat;\n  top: 50%;\n  left: 50%;\n  transform: translate(-50%, -50%);\n  opacity: 0.4;\n  animation: ", " 7s ease-in-out infinite;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n        0% {\n            transform: translate(-50%, -50%) translateY(0px)\n    }\n    \n        50% {\n            transform: translate(-50%, -50%) translateY(-15px)\n    }\n    \n        100% {\n            transform: translate(-50%, -50%) translateY(0px)\n    }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 100%;\n  height: 100%;\n  z-index: 9999;\n  background: white;\n  position: absolute;\n  top: 0;\n  left: 0;\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  flex-direction: column;\n  text-align: center;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var Container = styled__default['default'].div(_templateObject());
var IMAGE_SRC$1 = "https://d27rt3a60hh1lx.cloudfront.net/images/spaces-logo.png";

var _float = core.keyframes(_templateObject2());

var Image$1 = styled__default['default'].div(_templateObject3(), IMAGE_SRC$1, _float);
var INVALID_KEYWORDS = ["FBAN", "FBAV", "Instagram"];

var BrowserChecker = function BrowserChecker(props) {
  var children = props.children;

  var _useState = react.useState(true),
      valid = _useState[0],
      setValid = _useState[1];

  react.useEffect(function () {
    var ua = navigator.userAgent || navigator.vendor || "";
    var valid = INVALID_KEYWORDS.filter(function (val) {
      return ua.includes(val);
    });
    setValid(valid.length === 0);
  }, []);

  if (valid) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, children);
  }

  return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(Image$1, null), "This browser cannot run our application.", /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("br", null), "Click those three dots at the top right", /*#__PURE__*/React.createElement("br", null), "to open in Safari or Chrome.");
};

function _templateObject$1() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: fixed;\n  top: 50vh;\n  left: 50vw;\n  z-index: 1;\n  mix-blend-mode: difference;\n\n  &::before {\n    content: \"\";\n    display: block;\n    height: 11px;\n    border-left: 1px solid #fff;\n    margin-bottom: -6px;\n  }\n\n  &::after {\n    content: \"\";\n    display: block;\n    width: 11px;\n    border-top: 1px solid #fff;\n    margin-left: -5px;\n  }\n"]);

  _templateObject$1 = function _templateObject() {
    return data;
  };

  return data;
}
var Element = styled__default['default'].div(_templateObject$1());

var Crosshair = function Crosshair() {
  return /*#__PURE__*/React.createElement(Element, null);
};

var DefaultTouch = {
  pos: new THREE.Vector2(0, 0),
  id: -1
};
var DRAG_SENSITIVITY = new THREE.Vector2(0.7, 0.7);
/**
 * TouchFPSCamera controls the camera rotation by detecting
 * touch drag on the screen. Unlike MouseFPSCamera, this component
 * does not have a way to pause, that must be done externally.
 *
 * @param props
 * @constructor
 */

var TouchFPSCamera = function TouchFPSCamera(props) {
  var quaternion = props.quaternion,
      position = props.position;
  var touchStartPos = react.useRef(DefaultTouch);
  var originEuler = react.useRef(new THREE.Euler(0, 0, 0, "YXZ"));

  var _useThree = reactThreeFiber.useThree(),
      camera = _useThree.camera;

  reactThreeFiber.useFrame(function () {
    if (position.current) {
      var _camera$position;

      var _position$current = position.current,
          pX = _position$current.x,
          pY = _position$current.y,
          pZ = _position$current.z;
      camera == null ? void 0 : (_camera$position = camera.position) == null ? void 0 : _camera$position.set(pX, pY, pZ);
    }
  });

  var getNewEuler = function getNewEuler(dragX, dragY) {
    var newEuler = originEuler.current.clone();
    var moveX = dragX - touchStartPos.current.pos.x;
    var moveY = dragY - touchStartPos.current.pos.y;
    newEuler.setFromQuaternion(camera.quaternion);
    newEuler.y = originEuler.current.y - moveX * DRAG_SENSITIVITY.x / 100;
    newEuler.x = originEuler.current.x - moveY * DRAG_SENSITIVITY.y / 100;
    newEuler.x = Math.max(-Math.PI / 2, Math.min(Math.PI / 2, newEuler.x));
    return newEuler;
  };

  var tappedNipple = function tappedNipple(ev) {
    // get the relevant touched element (casted as an Element)
    var ele = ev.touches[ev.touches.length - 1].target;
    return ele.classList.contains("nipple-container") || ele.classList.contains("front") || ele.classList.contains("back");
  };

  var getCurrentTouch = function getCurrentTouch(touches) {
    var len = touches.length;

    for (var i = 0; i < len; i++) {
      if (touchStartPos.current.id === touches[i].identifier) {
        return touches[i];
      }
    }

    return undefined;
  }; // touch move scripts


  var onTouchStart = function onTouchStart(ev) {
    if (touchStartPos.current.id !== -1) {
      return;
    }

    if (tappedNipple(ev)) {
      touchStartPos.current = DefaultTouch;
      return;
    } // get last in list (most recent touch) to not confuse with movement


    var touchIndex = ev.touches.length - 1;
    var _ev$touches$touchInde = ev.touches[touchIndex],
        clientX = _ev$touches$touchInde.clientX,
        clientY = _ev$touches$touchInde.clientY,
        id = _ev$touches$touchInde.identifier;
    touchStartPos.current = {
      pos: new THREE.Vector2(clientX, clientY),
      id: id
    };
  };

  var onTouchMove = function onTouchMove(ev) {
    var touch = getCurrentTouch(ev.touches);

    if (!touch) {
      return;
    }

    var clientX = touch.clientX,
        clientY = touch.clientY;
    var newEuler = getNewEuler(clientX, clientY);
    camera.quaternion.setFromEuler(newEuler);
    quaternion.current = camera.quaternion;
  };

  var onTouchEnd = function onTouchEnd(ev) {
    var touch = getCurrentTouch(ev.changedTouches);

    if (!touch) {
      return;
    }

    var clientX = touch.clientX,
        clientY = touch.clientY;
    originEuler.current = getNewEuler(clientX, clientY);
    touchStartPos.current.id = -1;
  };

  react.useEffect(function () {
    camera == null ? void 0 : camera.lookAt(0, 2, 0);
    quaternion.current = camera.quaternion;
  }, []);
  react.useEffect(function () {
    document.addEventListener("touchstart", onTouchStart);
    document.addEventListener("touchmove", onTouchMove);
    document.addEventListener("touchend", onTouchEnd);
    return function () {
      document.removeEventListener("touchstart", onTouchStart);
      document.removeEventListener("touchmove", onTouchMove);
      document.removeEventListener("touchend", onTouchEnd);
    };
  }, []); // @ts-ignore

  return /*#__PURE__*/React.createElement(React.Fragment, null);
};

/**
 * NippleMovement gives the player a direction to move by taking
 * input from a nipple (joystick).
 *
 * Direction is stored as a Vector3 with the following format
 *    x: left/right movement, + for right
 *    y: forward/back movement, + for forwards
 *    z: up/down movement, + for up
 *
 * @param props
 * @constructor
 */
var NippleMovement = function NippleMovement(props) {
  var direction = props.direction,
      nippleContainer = props.nippleContainer;
  var nipple = react.useRef();

  var _useEnvironment = useEnvironment(),
      container = _useEnvironment.container;

  react.useEffect(function () {
    if (container) {
      nippleContainer.current = document.createElement("divs");
      nippleContainer.current.style.position = "fixed";
      nippleContainer.current.style.left = "0";
      nippleContainer.current.style.bottom = "0";
      nippleContainer.current.style.width = "40%";
      nippleContainer.current.style.maxWidth = "160px";
      nippleContainer.current.style.height = "25%";
      nippleContainer.current.style.height = "160px";
      nippleContainer.current.style.zIndex = "5"; // add class identifier to nippleContainer to identify touchEvents

      nippleContainer.current.classList.add("nipple-container");
      container.appendChild(nippleContainer.current);
      nipple.current = nipplejs__default['default'].create({
        zone: nippleContainer.current,
        mode: "static",
        position: {
          left: "50%",
          top: "50%"
        },
        color: "gray",
        size: 120,
        restOpacity: 0.5
      });
      nipple.current.on("move", function (evt, data) {
        direction.current.set(data.vector.x, -data.vector.y, 0);
      });
      nipple.current.on("end", function () {
        direction.current.set(0, 0, 0);
      });
      return function () {
        if (nipple.current) {
          nipple.current.destroy();
        }
      };
    }
  }, []);
  return /*#__PURE__*/React.createElement(React.Fragment, null);
};

var MobileControls = function MobileControls(props) {
  var direction = props.direction,
      quaternion = props.quaternion,
      position = props.position;
  var nippleContainer = react.useRef(null);
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(NippleMovement, {
    direction: direction,
    nippleContainer: nippleContainer
  }), /*#__PURE__*/React.createElement(TouchFPSCamera, {
    quaternion: quaternion,
    position: position,
    nippleContainer: nippleContainer
  }));
};

/**
 * KeyboardMovement gives the player a direction to move by taking
 * input from any source (currently keyboard) and calculating
 * relative direction.
 *
 * Direction is stored as a Vector3 with the following format
 *    x: left/right movement, + for right
 *    y: forward/back movement, + for forwards
 *    z: up/down movement, + for up
 *
 * @param props
 * @constructor
 */
var KeyboardMovement = function KeyboardMovement(props) {
  var direction = props.direction;
  var pressedKeys = react.useRef([false, false, false, false]); // key events

  var calcDirection = function calcDirection() {
    var press = pressedKeys.current; // [w, a, s, d]

    var yAxis = -1 * Number(press[0]) + Number(press[2]);
    var xAxis = -1 * Number(press[1]) + Number(press[3]);
    return new THREE.Vector3(xAxis, yAxis, 0);
  };

  var onKeyDown = function onKeyDown(ev) {
    if (ev.key === "w" || ev.key === "W" || ev.key === "ArrowUp") {
      pressedKeys.current[0] = true;
    }

    if (ev.key === "a" || ev.key === "A" || ev.key === "ArrowLeft") {
      pressedKeys.current[1] = true;
    }

    if (ev.key === "s" || ev.key === "S" || ev.key === "ArrowDown") {
      pressedKeys.current[2] = true;
    }

    if (ev.key === "d" || ev.key === "D" || ev.key === "ArrowRight") {
      pressedKeys.current[3] = true;
    }

    direction.current = calcDirection();
  };

  var onKeyUp = function onKeyUp(ev) {
    if (ev.key === "w" || ev.key === "W" || ev.key === "ArrowUp") {
      pressedKeys.current[0] = false;
    }

    if (ev.key === "a" || ev.key === "A" || ev.key === "ArrowLeft") {
      pressedKeys.current[1] = false;
    }

    if (ev.key === "s" || ev.key === "S" || ev.key === "ArrowDown") {
      pressedKeys.current[2] = false;
    }

    if (ev.key === "d" || ev.key === "D" || ev.key === "ArrowRight") {
      pressedKeys.current[3] = false;
    }

    direction.current = calcDirection();
  };

  react.useEffect(function () {
    document.addEventListener("keydown", onKeyDown);
    document.addEventListener("keyup", onKeyUp);
    return function () {
      document.removeEventListener("keydown", onKeyDown);
      document.removeEventListener("keyup", onKeyUp);
    };
  }, [onKeyUp, onKeyDown]);
  return /*#__PURE__*/React.createElement(React.Fragment, null);
};

reactThreeFiber.extend({
  PointerLockControls: PointerLockControls.PointerLockControls
});

/**
 * MouseFPSCamera controls the camera rotation, set up as a first
 * person view. It takes care of user input to set the camera rotation
 * and passed the quaternion of the camera up. Its position is attached
 * to the passed position, probably the player's position.
 *
 * @param props
 * @constructor
 */
var MouseFPSCamera = function MouseFPSCamera(props) {
  var quaternion = props.quaternion,
      position = props.position;

  var _useEnvironment = useEnvironment(),
      setPaused = _useEnvironment.setPaused,
      paused = _useEnvironment.paused,
      addEvent = _useEnvironment.addEvent;

  var _useThree = reactThreeFiber.useThree(),
      camera = _useThree.camera,
      gl = _useThree.gl;

  var controls = react.useRef();
  reactThreeFiber.useFrame(function () {
    var _controls$current;

    if ((controls == null ? void 0 : (_controls$current = controls.current) == null ? void 0 : _controls$current.isLocked) === true) {
      var lookAt = new THREE.Vector3(0, 0, -1);
      lookAt.applyQuaternion(camera.quaternion);
      lookAt.multiply(new THREE.Vector3(1, 0, 1)).normalize();
      quaternion.current = camera.quaternion;
    }

    if (position.current) {
      var _camera$position;

      var _position$current = position.current,
          pX = _position$current.x,
          pY = _position$current.y,
          pZ = _position$current.z;
      camera == null ? void 0 : (_camera$position = camera.position) == null ? void 0 : _camera$position.set(pX, pY, pZ);
    }
  }); // pointer locking events

  var onLock = function onLock() {
    // dont't need to set if it's already set
    if (paused) {
      setPaused(false);
    }
  };

  var onUnlock = function onUnlock() {
    // dont't need to set if it's already set
    if (!paused) {
      setPaused(true);
    }
  };

  var lockCamera = function lockCamera() {
    var _controls$current2;

    return controls == null ? void 0 : (_controls$current2 = controls.current) == null ? void 0 : _controls$current2.lock();
  };

  var unlockCamera = function unlockCamera() {
    var _controls$current3;

    return controls == null ? void 0 : (_controls$current3 = controls.current) == null ? void 0 : _controls$current3.unlock();
  };

  react.useEffect(function () {
    // initial camera rotation
    if (camera && quaternion) {
      camera == null ? void 0 : camera.lookAt(0, 2, 0);
      quaternion.current = camera.quaternion;
    } // lock and unlock controls based on set paused value


    addEvent("paused", function (pausedVal, overlayVal) {
      if (pausedVal) {
        unlockCamera();
      } else {
        lockCamera();
      }
    });
  }, []);
  react.useEffect(function () {
    var _controls$current5, _controls$current6;

    setTimeout(function () {
      var _controls$current4;

      if (!(controls == null ? void 0 : (_controls$current4 = controls.current) == null ? void 0 : _controls$current4.isLocked) && !paused) {
        setPaused(true);
      }
    }, 250);
    controls == null ? void 0 : (_controls$current5 = controls.current) == null ? void 0 : _controls$current5.addEventListener("lock", onLock);
    controls == null ? void 0 : (_controls$current6 = controls.current) == null ? void 0 : _controls$current6.addEventListener("unlock", onUnlock);
    return function () {
      var _controls$current7, _controls$current8;

      controls == null ? void 0 : (_controls$current7 = controls.current) == null ? void 0 : _controls$current7.removeEventListener("lock", onLock);
      controls == null ? void 0 : (_controls$current8 = controls.current) == null ? void 0 : _controls$current8.removeEventListener("unlock", onUnlock);
    };
  }, [paused, controls]); // @ts-ignore

  return /*#__PURE__*/React.createElement("pointerLockControls", {
    ref: controls,
    args: [camera, gl.domElement]
  });
};

var DesktopControls = function DesktopControls(props) {
  var direction = props.direction,
      quaternion = props.quaternion,
      position = props.position;
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(KeyboardMovement, {
    direction: direction
  }), /*#__PURE__*/React.createElement(MouseFPSCamera, {
    quaternion: quaternion,
    position: position
  }));
};

function createPlayerRef(bodyApi, position, velocity, lockControls, raycaster) {
  var posVec = {
    set: function set(vec) {
      return bodyApi.position.set(vec.x, vec.y, vec.z);
    },
    get: function get() {
      return position.current;
    }
  };
  var velVec = {
    set: function set(vec) {
      return bodyApi.velocity.set(vec.x, vec.y, vec.z);
    },
    get: function get() {
      return velocity.current;
    }
  };
  var controls = {
    lock: function lock() {
      return lockControls.current = true;
    },
    unlock: function unlock() {
      return lockControls.current = false;
    },
    isLocked: function isLocked() {
      return lockControls.current;
    }
  };
  var playerRef = {
    position: posVec,
    velocity: velVec,
    controls: controls,
    raycaster: raycaster.current
  };
  return playerRef;
}

var VELOCITY_FACTOR = 250;
var SHOW_PLAYER_HITBOX = false;

/**
 * Player represents a physics-enabled player in the environment, complete with a
 * control scheme and a physical representation that interacts with other physics-
 * enabled objects.
 *
 * There should only be one player per environment.
 *
 * @constructor
 */
var Player = function Player(props) {
  var _props$initPos = props.initPos,
      initPos = _props$initPos === void 0 ? new THREE.Vector3(0, 1, 0) : _props$initPos,
      _props$initRot = props.initRot,
      initRot = _props$initRot === void 0 ? 0 : _props$initRot;

  var _useThree = reactThreeFiber.useThree(),
      camera = _useThree.camera;

  var _useEnvironment = useEnvironment(),
      paused = _useEnvironment.paused,
      setPlayer = _useEnvironment.setPlayer; // physical body


  var _useSphere = cannon.useSphere(function () {
    return {
      mass: 500,
      position: initPos.toArray(),
      args: 1,
      fixedRotation: true
    };
  }),
      bodyRef = _useSphere[0],
      bodyApi = _useSphere[1]; // producer


  var prevTime = react.useRef(performance.now());
  var position = react.useRef(new THREE.Vector3(0, 0, 0));
  var velocity = react.useRef(new THREE.Vector3(0, 0, 0));
  var lockControls = react.useRef(false);
  var raycaster = react.useRef(new THREE.Raycaster(new THREE.Vector3(), new THREE.Vector3(), 0, 3)); // consumer

  var direction = react.useRef(new THREE.Vector3());
  var quaternion = react.useRef(new THREE.Quaternion(0, 0, 0, 0)); // rad on y axis
  // setup player

  react.useEffect(function () {
    // store position and velocity
    bodyApi.position.subscribe(function (p) {
      position.current.set(p[0], p[1], p[2]);
    });
    bodyApi.velocity.subscribe(function (v) {
      return velocity.current.set(v[0], v[1], v[2]);
    });
    var xLook = initPos.x + 100 * Math.cos(initRot);
    var zLook = initPos.z + 100 * Math.sin(initRot);
    camera == null ? void 0 : camera.lookAt(xLook, initPos.y, zLook);
    setPlayer(createPlayerRef(bodyApi, position, velocity, lockControls, raycaster));
  }, []); // update player every frame

  reactThreeFiber.useFrame(function () {
    var time = performance.now(); // update raycaster

    if (position.current && quaternion.current) {
      raycaster.current.ray.origin.copy(position.current);
      var lookAt = new THREE.Vector3(0, 0, -1);
      lookAt.applyQuaternion(quaternion.current);
      raycaster.current.ray.direction.copy(lookAt);
    }

    if (paused) {
      // stop player from moving when paused
      bodyApi == null ? void 0 : bodyApi.velocity.set(0, 0, 0);
    } else {
      // get time since last computation
      var delta = (time - prevTime.current) / 1000; // get forward/back movement and left/right movement velocities

      var inputVelocity = new THREE.Vector3(0, 0, 0);

      if (!lockControls.current) {
        inputVelocity.x = direction.current.x * delta * 0.75 * VELOCITY_FACTOR;
        inputVelocity.z = direction.current.y * delta * VELOCITY_FACTOR;
      } // apply quaternion to get adjusted direction based on camera


      var moveQuaternion = quaternion.current.clone();
      moveQuaternion.x = 0;
      moveQuaternion.z = 0;
      inputVelocity.applyQuaternion(moveQuaternion); // keep y velocity intact and update velocity

      inputVelocity.add(new THREE.Vector3(0, velocity.current.y, 0));
      bodyApi == null ? void 0 : bodyApi.velocity.set(inputVelocity.x, inputVelocity.y, inputVelocity.z);
    }

    prevTime.current = time;
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, reactDeviceDetect.isMobile ? /*#__PURE__*/React.createElement(MobileControls, {
    quaternion: quaternion,
    position: position,
    direction: direction
  }) : /*#__PURE__*/React.createElement(DesktopControls, {
    quaternion: quaternion,
    position: position,
    direction: direction
  }), /*#__PURE__*/React.createElement("mesh", {
    ref: bodyRef,
    name: "player"
  }, SHOW_PLAYER_HITBOX ));
};

function _templateObject$2() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 100%;\n  height: 100%;\n  position: absolute;\n  top: 0;\n  left: 0;\n  z-index: 200;\n  background: white;\n  transition: opacity 0.5s ease;\n  transition-delay: 0.25s;\n  opacity: ", ";\n  pointer-events: ", ";\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  flex-direction: column;\n"]);

  _templateObject$2 = function _templateObject() {
    return data;
  };

  return data;
}
var Container$1 = styled__default['default'].div(_templateObject$2(), function (props) {
  return props.finished ? 0 : 1;
}, function (props) {
  return props.finished ? "none" : "all";
});

var LoadingScreen = function LoadingScreen() {
  var progress = useControlledProgress();
  return /*#__PURE__*/React.createElement(Container$1, {
    finished: progress == 100,
    landing: false
  }, progress, "%");
};

reactThreeFiber.extend({
  EffectComposer: EffectComposer.EffectComposer,
  RenderPass: RenderPass.RenderPass,
  GlitchPass: GlitchPass.GlitchPass,
  ShaderPass: ShaderPass.ShaderPass,
  BloomPass: BloomPass.BloomPass
});
var RealisticEffects = function RealisticEffects() {
  var _useThree = reactThreeFiber.useThree(),
      gl = _useThree.gl,
      scene = _useThree.scene,
      camera = _useThree.camera,
      size = _useThree.size;

  var gammaCorrection = new ShaderPass.ShaderPass(GammaCorrectionShader.GammaCorrectionShader);
  var fxaaPass = new ShaderPass.ShaderPass(FXAAShader.FXAAShader);
  var pixelRatio = window ? window.devicePixelRatio : 2; // @ts-ignore

  fxaaPass.material.uniforms["resolution"].value.x = 1 / (window.innerWidth * pixelRatio); // @ts-ignore

  fxaaPass.material.uniforms["resolution"].value.y = 1 / (window.innerHeight * pixelRatio);
  var composer = react.useRef();
  react.useEffect(function () {
    var _composer$current;

    return void (composer == null ? void 0 : (_composer$current = composer.current) == null ? void 0 : _composer$current.setSize(size.width, size.height));
  }, [size]);
  react.useEffect(function () {
    var _composer$current2, _composer$current3;

    gammaCorrection.renderToScreen = true;
    composer == null ? void 0 : (_composer$current2 = composer.current) == null ? void 0 : _composer$current2.addPass(gammaCorrection);
    fxaaPass.renderToScreen = true;
    composer == null ? void 0 : (_composer$current3 = composer.current) == null ? void 0 : _composer$current3.addPass(fxaaPass);
  }, []);
  reactThreeFiber.useFrame(function () {
    var _composer$current4;

    composer == null ? void 0 : (_composer$current4 = composer.current) == null ? void 0 : _composer$current4.render();
  }, 1);
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("effectComposer", {
    ref: composer,
    args: [gl]
  }, /*#__PURE__*/React.createElement("renderPass", {
    attachArray: "passes",
    args: [scene, camera]
  })));
};

function _templateObject9() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 100%;\n  height: auto;\n  margin: 10px 0;\n  font-family: \"Space Mono\", monospace;\n  font-size: 0.7em;\n  text-align: center;\n  display: flex;\n  flex-direction: column;\n  justify-content: center;\n"]);

  _templateObject9 = function _templateObject9() {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  font-size: 2em;\n  text-align: center;\n  margin-bottom: 0em;\n  line-height: 1em;\n"]);

  _templateObject8 = function _templateObject8() {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  margin-top: 8%;\n  width: auto;\n  display: flex;\n  flex-direction: column;\n  align-items: center;\n  font-family: \"Lato\", sans-serif;\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: absolute;\n  top: 24px;\n  left: 60px;\n  width: auto;\n  height: auto;\n  color: white;\n  cursor: pointer;\n  transition: opacity 0.15s linear;\n  font-size: 0.6em;\n  line-height: 1em;\n  :hover {\n    opacity: 0.5;\n  }\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: absolute;\n  top: 24px;\n  right: 60px;\n  font-size: 0.6em;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 90%;\n  max-width: 400px;\n  height: auto;\n  cursor: pointer;\n  text-align: center;\n  font-size: 1.3em;\n  font-family: \"Lato\", sans-serif;\n  transition: opacity 0.15s linear;\n  margin-top: 20px;\n  background: white;\n  line-height: 1em;\n  padding: 12px 0;\n  border-radius: 10px;\n  :hover {\n    opacity: 0.5;\n  }\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3$1() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 90%;\n  max-width: 400px;\n  height: 91vw;\n  max-height: 400px;\n  padding: 20px 20px;\n  color: white;\n  display: flex;\n  flex-direction: column;\n  align-items: center;\n  justify-content: center;\n  position: relative;\n  border-radius: 3%;\n  background-image: url(\"https://spaces-gallery-assets.s3-us-west-1.amazonaws.com/images/pauseMenuBg.png\");\n  background-position: center;\n  background-size: cover;\n"]);

  _templateObject3$1 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$1() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 100%;\n  height: 100%;\n  position: absolute;\n  top: 0;\n  left: 0;\n  z-index: -1;\n"]);

  _templateObject2$1 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$3() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 100%;\n  height: 100%;\n  position: absolute;\n  top: 0;\n  left: 0;\n  z-index: 100;\n  transition: opacity 0.25s ease;\n  background: rgba(0, 0, 0, 0.5);\n  display: flex;\n  justify-content: center;\n  flex-direction: column;\n  align-items: center;\n  opacity: ", ";\n  pointer-events: ", ";\n"]);

  _templateObject$3 = function _templateObject() {
    return data;
  };

  return data;
}
var NEXT_PUBLIC_VERSION = process.env.NEXT_PUBLIC_VERSION;
var Container$2 = styled__default['default'].div(_templateObject$3(), function (props) {
  return props.paused ? 1 : 0;
}, function (props) {
  return props.paused ? "all" : "none";
});
var ClickContainer = styled__default['default'].div(_templateObject2$1());
var Window = styled__default['default'].div(_templateObject3$1());
var Continue = styled__default['default'].div(_templateObject4());
var Version = styled__default['default'].a(_templateObject5());
var Instagram = styled__default['default'].div(_templateObject6());
var Header = styled__default['default'].div(_templateObject7());
var Title = styled__default['default'].div(_templateObject8());
var Text$1 = styled__default['default'].div(_templateObject9());

var DesktopPause = function DesktopPause() {
  var _useEnvironment = useEnvironment(),
      paused = _useEnvironment.paused,
      overlay = _useEnvironment.overlay,
      setPaused = _useEnvironment.setPaused;

  var closeOverlay = function closeOverlay() {
    return setPaused(false);
  };

  if (overlay) {
    return null;
  }

  return /*#__PURE__*/React.createElement(Container$2, {
    paused: paused
  }, /*#__PURE__*/React.createElement(ClickContainer, {
    onClick: closeOverlay
  }), /*#__PURE__*/React.createElement(Window, null, /*#__PURE__*/React.createElement(Version, null, "v ", NEXT_PUBLIC_VERSION), /*#__PURE__*/React.createElement(Instagram, {
    onClick: function onClick() {
      window.open("https://www.instagram.com/spaces3.0");
    }
  }, "@spaces3.0"), /*#__PURE__*/React.createElement(Header, null, /*#__PURE__*/React.createElement(Title, null, "SPACES")), /*#__PURE__*/React.createElement(Text$1, null, /*#__PURE__*/React.createElement("p", null, "Move around: ", reactDeviceDetect.isMobile ? "Joystick" : "W/A/S/D"), /*#__PURE__*/React.createElement("p", null, "Look around: ", reactDeviceDetect.isMobile ? "Drag" : "Mouse"), /*#__PURE__*/React.createElement("p", null, "Pause: ", reactDeviceDetect.isMobile ? "Menu Button" : "Esc"))), /*#__PURE__*/React.createElement(Continue, {
    onClick: closeOverlay
  }, "continue"));
};

function _templateObject3$2() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 60px;\n  height: 60px;\n  margin: 10px;\n  border-radius: 50%;\n  text-align: center;\n  background: url(\"https://spaces-gallery-assets.s3-us-west-1.amazonaws.com/images/spacesLogoReal.png\")\n    center center no-repeat;\n  background-size: cover;\n"]);

  _templateObject3$2 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$2() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 100%;\n  height: 100%;\n  position: absolute;\n  top: 0;\n  right: 0;\n  z-index: 10;\n"]);

  _templateObject2$2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$4() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: absolute;\n  top: 0;\n  right: 0;\n  z-index: 9;\n"]);

  _templateObject$4 = function _templateObject() {
    return data;
  };

  return data;
}
var Container$3 = styled__default['default'].div(_templateObject$4());
var ClickContainer$1 = styled__default['default'].div(_templateObject2$2());
var Window$1 = styled__default['default'].div(_templateObject3$2());

var MobilePause = function MobilePause() {
  var _useEnvironment = useEnvironment(),
      paused = _useEnvironment.paused,
      setPaused = _useEnvironment.setPaused;

  var togglePause = function togglePause() {
    return setPaused(!paused);
  };

  return /*#__PURE__*/React.createElement(Container$3, {
    paused: paused
  }, /*#__PURE__*/React.createElement(ClickContainer$1, {
    onClick: togglePause
  }), !paused ? /*#__PURE__*/React.createElement(Window$1, null) : null);
};

function _templateObject$5() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  @font-face {\n    font-family: \"MyriadPro\";\n    src: url(\"https://d27rt3a60hh1lx.cloudfront.net/fonts/Myriad_Pro_Light.otf\");\n  }\n\n  @font-face {\n    font-family: \"Lomino\";\n    src: url(\"https://d27rt3a60hh1lx.cloudfront.net/fonts/LominoUI_A_Rg.ttf\");\n  }\n\n  html {\n    position: fixed;\n    height: 100%;\n    overflow: hidden;\n  }\n\n  body {\n    margin: 0;\n    width: 100vw;\n    height: 100vh;\n    user-select: none;\n    overflow-y: auto;\n    overflow-x: hidden;\n    touch-action: pan-x pan-y;\n    -webkit-overflow-scrolling: touch;\n    font-family: \"MyriadPro\", sans-serif;\n    font-size: 27px;\n    @media screen and (max-width: 500px) {\n      font-size: 24px;\n    }\n  }\n"]);

  _templateObject$5 = function _templateObject() {
    return data;
  };

  return data;
}
var globalStyles = core.css(_templateObject$5());
var GlobalStyles = (function () {
  return /*#__PURE__*/React.createElement(core.Global, {
    styles: globalStyles
  });
});

function _templateObject$6() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: absolute;\n  width: 100%;\n  height: 100%;\n  overflow: hidden;\n\n  canvas {\n    position: absolute;\n    top: 0;\n    right: 0;\n    bottom: 0;\n    left: 0;\n    outline: 0;\n  }\n"]);

  _templateObject$6 = function _templateObject() {
    return data;
  };

  return data;
}
var Container$4 = styled__default['default'].div(_templateObject$6());
var defaultCanvasProps = {
  shadowMap: true,
  gl: {
    alpha: false
  },
  camera: {
    position: [0, 2, 0],
    near: 0.01,
    far: 100
  }
};
var defaultPhysicsProps = {
  iterations: 20,
  size: 10,
  allowSleep: false,
  defaultContactMaterial: {
    friction: 0
  }
};

/**
 *
 * Standard environment should be your first choice for an environment
 *
 * Player Type: First Person w/ WASD, Joystick controls
 * Physics: Enabled with default y=0 floor plane
 * Loading Screen: Spaces Edition
 * Pause Menu: Spaces Edition
 *
 */
var StandardEnvironment = function StandardEnvironment(props) {
  var children = props.children,
      canvasProps = props.canvasProps,
      physicsProps = props.physicsProps,
      player = props.player,
      disableGround = props.disableGround,
      disableEffects = props.disableEffects;
  var state = useEnvironmentState();
  return /*#__PURE__*/React.createElement(BrowserChecker, null, /*#__PURE__*/React.createElement(GlobalStyles, null), /*#__PURE__*/React.createElement(Container$4, {
    ref: state.containerRef
  }, /*#__PURE__*/React.createElement(reactThreeFiber.Canvas, _extends__default['default']({}, defaultCanvasProps, canvasProps), /*#__PURE__*/React.createElement(cannon.Physics, _extends__default['default']({}, defaultPhysicsProps, physicsProps), /*#__PURE__*/React.createElement(environmentStateContext.Provider, {
    value: state
  }, /*#__PURE__*/React.createElement(Player, {
    initPos: player == null ? void 0 : player.pos,
    initRot: player == null ? void 0 : player.rot
  }), !disableGround && /*#__PURE__*/React.createElement(InfinitePlane, {
    height: -0.001
  }), !disableEffects && /*#__PURE__*/React.createElement(RealisticEffects, null), children))), /*#__PURE__*/React.createElement(environmentStateContext.Provider, {
    value: state
  }, /*#__PURE__*/React.createElement(LoadingScreen, null), /*#__PURE__*/React.createElement(DesktopPause, null), reactDeviceDetect.isMobile && /*#__PURE__*/React.createElement(MobilePause, null), /*#__PURE__*/React.createElement(Crosshair, null))));
};

function _templateObject4$1() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: fixed;\n  bottom: 5vh;\n  color: black;\n  font-size: 1.5rem;\n  cursor: pointer;\n  :hover {\n    text-decoration: underline;\n  }\n  transition: opacity 1s ease;\n  transition-delay: 2s;\n\n  @media screen and (max-width: 650px) {\n    font-size: 1rem;\n  }\n"]);

  _templateObject4$1 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3$3() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  padding: 0.5rem 0.5rem 0.5rem 0.5rem;\n  background-image: linear-gradient(120deg, magenta, #5b328d);\n  background-size: 120% 120%;\n  background-position: top left;\n  border-radius: 5px;\n  transition: opacity 1s ease, background-position 0.3s linear;\n  margin-bottom: 20px;\n\n  &:hover,\n  &:active {\n    background-position: bottom right;\n  }\n"]);

  _templateObject3$3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$3() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  color: white;\n  font-size: 4rem;\n  font-weight: bold;\n  outline: none;\n  height: auto;\n  width: auto;\n  padding: 20px 20px 20px 20px;\n  transition: opacity 1s ease;\n  transition-delay: 0.5s;\n  border-radius: 5px;\n  background-color: transparent;\n  cursor: pointer;\n  border: none;\n  font-family: \"Lato\", sans-serif;\n\n  @media screen and (max-width: 650px) {\n    font-size: 2rem;\n  }\n"]);

  _templateObject2$3 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$7() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 100%;\n  height: 100%;\n  position: absolute;\n  top: 0;\n  left: 0;\n  z-index: 200;\n  background: white;\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  flex-direction: column;\n  font-family: \"Lato\", sans-serif;\n\n  transition: opacity 0.7s ease;\n  ", "\n  ", "\n"]);

  _templateObject$7 = function _templateObject() {
    return data;
  };

  return data;
}
var Container$5 = styled__default$1['default'].div(_templateObject$7(), function (props) {
  return props.finished && "opacity: 0;";
}, function (props) {
  return props.finished && "pointer-events: none;";
});
var Begin = styled__default$1['default'].button(_templateObject2$3());
var BeginBackground = styled__default$1['default'].div(_templateObject3$3());
var AlternateStart = styled__default$1['default'].div(_templateObject4$1());

var Transition = function Transition(props) {
  var start = props.start,
      beginExperience = props.beginExperience,
      selfExplore = props.selfExplore;
  return /*#__PURE__*/React.createElement(Container$5, {
    finished: start
  }, /*#__PURE__*/React.createElement(BeginBackground, null, /*#__PURE__*/React.createElement(Begin, {
    onClick: beginExperience
  }, "Experience")), /*#__PURE__*/React.createElement(AlternateStart, {
    onClick: selfExplore
  }, "Or, Explore on Your Own"));
};

function _templateObject2$4() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: absolute;\n  top: 50%;\n  left: 50%;\n  z-index: 2;\n  perspective: 500px;\n  -moz-perspective: 500px;\n  -webkit-perspective: 500px;\n\n  & > i {\n    display: block;\n    position: absolute;\n    width: 8px;\n    height: 8px;\n    border-radius: 8px;\n    opacity: 0;\n    background: rgba(255, 0, 255, 0.5);\n    box-shadow: 0px 0px 10px white;\n    animation-name: ", ";\n    animation-duration: 3s;\n    animation-iteration-count: infinite;\n    animation-timing-function: ease-in-out;\n  }\n\n  & i:nth-of-type(1) {\n    transform: rotate(11.6129deg) translate3d(80px, 0, 0);\n    animation-delay: 0.04839s;\n  }\n\n  i:nth-of-type(2) {\n    transform: rotate(23.22581deg) translate3d(80px, 0, 0);\n    animation-delay: 0.09677s;\n  }\n\n  i:nth-of-type(3) {\n    transform: rotate(34.83871deg) translate3d(80px, 0, 0);\n    animation-delay: 0.14516s;\n  }\n\n  i:nth-of-type(4) {\n    transform: rotate(46.45161deg) translate3d(80px, 0, 0);\n    animation-delay: 0.19355s;\n  }\n\n  i:nth-of-type(5) {\n    transform: rotate(58.06452deg) translate3d(80px, 0, 0);\n    animation-delay: 0.24194s;\n  }\n\n  i:nth-of-type(6) {\n    transform: rotate(69.67742deg) translate3d(80px, 0, 0);\n    animation-delay: 0.29032s;\n  }\n\n  i:nth-of-type(7) {\n    transform: rotate(81.29032deg) translate3d(80px, 0, 0);\n    animation-delay: 0.33871s;\n  }\n\n  i:nth-of-type(8) {\n    transform: rotate(92.90323deg) translate3d(80px, 0, 0);\n    animation-delay: 0.3871s;\n  }\n\n  i:nth-of-type(9) {\n    transform: rotate(104.51613deg) translate3d(80px, 0, 0);\n    animation-delay: 0.43548s;\n  }\n\n  i:nth-of-type(10) {\n    transform: rotate(116.12903deg) translate3d(80px, 0, 0);\n    animation-delay: 0.48387s;\n  }\n\n  i:nth-of-type(11) {\n    transform: rotate(127.74194deg) translate3d(80px, 0, 0);\n    animation-delay: 0.53226s;\n  }\n\n  i:nth-of-type(12) {\n    transform: rotate(139.35484deg) translate3d(80px, 0, 0);\n    animation-delay: 0.58065s;\n  }\n\n  i:nth-of-type(13) {\n    transform: rotate(150.96774deg) translate3d(80px, 0, 0);\n    animation-delay: 0.62903s;\n  }\n\n  i:nth-of-type(14) {\n    transform: rotate(162.58065deg) translate3d(80px, 0, 0);\n    animation-delay: 0.67742s;\n  }\n\n  i:nth-of-type(15) {\n    transform: rotate(174.19355deg) translate3d(80px, 0, 0);\n    animation-delay: 0.72581s;\n  }\n\n  i:nth-of-type(16) {\n    transform: rotate(185.80645deg) translate3d(80px, 0, 0);\n    animation-delay: 0.77419s;\n  }\n\n  i:nth-of-type(17) {\n    transform: rotate(197.41935deg) translate3d(80px, 0, 0);\n    animation-delay: 0.82258s;\n  }\n\n  i:nth-of-type(18) {\n    transform: rotate(209.03226deg) translate3d(80px, 0, 0);\n    animation-delay: 0.87097s;\n  }\n\n  i:nth-of-type(19) {\n    transform: rotate(220.64516deg) translate3d(80px, 0, 0);\n    animation-delay: 0.91935s;\n  }\n\n  i:nth-of-type(20) {\n    transform: rotate(232.25806deg) translate3d(80px, 0, 0);\n    animation-delay: 0.96774s;\n  }\n\n  i:nth-of-type(21) {\n    transform: rotate(243.87097deg) translate3d(80px, 0, 0);\n    animation-delay: 1.01613s;\n  }\n\n  i:nth-of-type(22) {\n    transform: rotate(255.48387deg) translate3d(80px, 0, 0);\n    animation-delay: 1.06452s;\n  }\n\n  i:nth-of-type(23) {\n    transform: rotate(267.09677deg) translate3d(80px, 0, 0);\n    animation-delay: 1.1129s;\n  }\n\n  i:nth-of-type(24) {\n    transform: rotate(278.70968deg) translate3d(80px, 0, 0);\n    animation-delay: 1.16129s;\n  }\n\n  i:nth-of-type(25) {\n    transform: rotate(290.32258deg) translate3d(80px, 0, 0);\n    animation-delay: 1.20968s;\n  }\n\n  i:nth-of-type(26) {\n    transform: rotate(301.93548deg) translate3d(80px, 0, 0);\n    animation-delay: 1.25806s;\n  }\n\n  i:nth-of-type(27) {\n    transform: rotate(313.54839deg) translate3d(80px, 0, 0);\n    animation-delay: 1.30645s;\n  }\n\n  i:nth-of-type(28) {\n    transform: rotate(325.16129deg) translate3d(80px, 0, 0);\n    animation-delay: 1.35484s;\n  }\n\n  i:nth-of-type(29) {\n    transform: rotate(336.77419deg) translate3d(80px, 0, 0);\n    animation-delay: 1.40323s;\n  }\n\n  i:nth-of-type(30) {\n    transform: rotate(348.3871deg) translate3d(80px, 0, 0);\n    animation-delay: 1.45161s;\n  }\n\n  i:nth-of-type(31) {\n    transform: rotate(360deg) translate3d(80px, 0, 0);\n    animation-delay: 1.5s;\n  }\n\n  i:nth-of-type(32) {\n    transform: rotate(371.6129deg) translate3d(80px, 0, 0);\n    animation-delay: 1.54839s;\n  }\n\n  i:nth-of-type(33) {\n    transform: rotate(383.22581deg) translate3d(80px, 0, 0);\n    animation-delay: 1.59677s;\n  }\n\n  i:nth-of-type(34) {\n    transform: rotate(394.83871deg) translate3d(80px, 0, 0);\n    animation-delay: 1.64516s;\n  }\n\n  i:nth-of-type(35) {\n    transform: rotate(406.45161deg) translate3d(80px, 0, 0);\n    animation-delay: 1.69355s;\n  }\n\n  i:nth-of-type(36) {\n    transform: rotate(418.06452deg) translate3d(80px, 0, 0);\n    animation-delay: 1.74194s;\n  }\n\n  i:nth-of-type(37) {\n    transform: rotate(429.67742deg) translate3d(80px, 0, 0);\n    animation-delay: 1.79032s;\n  }\n\n  i:nth-of-type(38) {\n    transform: rotate(441.29032deg) translate3d(80px, 0, 0);\n    animation-delay: 1.83871s;\n  }\n\n  i:nth-of-type(39) {\n    transform: rotate(452.90323deg) translate3d(80px, 0, 0);\n    animation-delay: 1.8871s;\n  }\n\n  i:nth-of-type(40) {\n    transform: rotate(464.51613deg) translate3d(80px, 0, 0);\n    animation-delay: 1.93548s;\n  }\n\n  i:nth-of-type(41) {\n    transform: rotate(476.12903deg) translate3d(80px, 0, 0);\n    animation-delay: 1.98387s;\n  }\n\n  i:nth-of-type(42) {\n    transform: rotate(487.74194deg) translate3d(80px, 0, 0);\n    animation-delay: 2.03226s;\n  }\n\n  i:nth-of-type(43) {\n    transform: rotate(499.35484deg) translate3d(80px, 0, 0);\n    animation-delay: 2.08065s;\n  }\n\n  i:nth-of-type(44) {\n    transform: rotate(510.96774deg) translate3d(80px, 0, 0);\n    animation-delay: 2.12903s;\n  }\n\n  i:nth-of-type(45) {\n    transform: rotate(522.58065deg) translate3d(80px, 0, 0);\n    animation-delay: 2.17742s;\n  }\n\n  i:nth-of-type(46) {\n    transform: rotate(534.19355deg) translate3d(80px, 0, 0);\n    animation-delay: 2.22581s;\n  }\n\n  i:nth-of-type(47) {\n    transform: rotate(545.80645deg) translate3d(80px, 0, 0);\n    animation-delay: 2.27419s;\n  }\n\n  i:nth-of-type(48) {\n    transform: rotate(557.41935deg) translate3d(80px, 0, 0);\n    animation-delay: 2.32258s;\n  }\n\n  i:nth-of-type(49) {\n    transform: rotate(569.03226deg) translate3d(80px, 0, 0);\n    animation-delay: 2.37097s;\n  }\n\n  i:nth-of-type(50) {\n    transform: rotate(580.64516deg) translate3d(80px, 0, 0);\n    animation-delay: 2.41935s;\n  }\n\n  i:nth-of-type(51) {\n    transform: rotate(592.25806deg) translate3d(80px, 0, 0);\n    animation-delay: 2.46774s;\n  }\n\n  i:nth-of-type(52) {\n    transform: rotate(603.87097deg) translate3d(80px, 0, 0);\n    animation-delay: 2.51613s;\n  }\n\n  i:nth-of-type(53) {\n    transform: rotate(615.48387deg) translate3d(80px, 0, 0);\n    animation-delay: 2.56452s;\n  }\n\n  i:nth-of-type(54) {\n    transform: rotate(627.09677deg) translate3d(80px, 0, 0);\n    animation-delay: 2.6129s;\n  }\n\n  i:nth-of-type(55) {\n    transform: rotate(638.70968deg) translate3d(80px, 0, 0);\n    animation-delay: 2.66129s;\n  }\n\n  i:nth-of-type(56) {\n    transform: rotate(650.32258deg) translate3d(80px, 0, 0);\n    animation-delay: 2.70968s;\n  }\n\n  i:nth-of-type(57) {\n    transform: rotate(661.93548deg) translate3d(80px, 0, 0);\n    animation-delay: 2.75806s;\n  }\n\n  i:nth-of-type(58) {\n    transform: rotate(673.54839deg) translate3d(80px, 0, 0);\n    animation-delay: 2.80645s;\n  }\n\n  i:nth-of-type(59) {\n    transform: rotate(685.16129deg) translate3d(80px, 0, 0);\n    animation-delay: 2.85484s;\n  }\n\n  i:nth-of-type(60) {\n    transform: rotate(696.77419deg) translate3d(80px, 0, 0);\n    animation-delay: 2.90323s;\n  }\n\n  i:nth-of-type(61) {\n    transform: rotate(708.3871deg) translate3d(80px, 0, 0);\n    animation-delay: 2.95161s;\n  }\n\n  i:nth-of-type(62) {\n    transform: rotate(720deg) translate3d(80px, 0, 0);\n    animation-delay: 3s;\n  }\n\n  animation-delay: 1s;\n  transition: transform 1.75s ease-out, opacity 1.75s ease-out;\n  ", ";\n  ", ";\n"]);

  _templateObject2$4 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$8() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n    from {\n        opacity: 0.0;\n    }\n    to {\n        opacity: 0.6;\n        transform: translate3d(-4px, -4px, 570px);\n    }\n"]);

  _templateObject$8 = function _templateObject() {
    return data;
  };

  return data;
}
var spin = core.keyframes(_templateObject$8());
var Wrapper = styled__default['default'].div(_templateObject2$4(), spin, function (props) {
  return props.finished && "transform: translate3d(0, 0, 500px)";
}, function (props) {
  return props.finished && "opacity: 0";
});

var SpinningLoading = function SpinningLoading(props) {
  var progress = props.progress;
  var listElements = [].concat(Array(62)).map(function (x, ind) {
    return /*#__PURE__*/React.createElement("i", {
      key: ind
    });
  });
  return /*#__PURE__*/React.createElement(Wrapper, {
    finished: progress === 100
  }, listElements);
};

function _templateObject4$2() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  animation: ", " 6s linear infinite;\n"]);

  _templateObject4$2 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3$4() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  transition: opacity 1s ease-out;\n  transition-delay: 1.25s;\n  ", "\n"]);

  _templateObject3$4 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$5() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  0%, 100% {\n    transform: rotate(-0.35deg) translateZ(66px);\n     opacity: 0.9;\n  }\n  25%, 75% {\n    transform: rotate(0.2deg) translateZ(-10px);\n     opacity: 0.3;\n  }\n  50% {\n    transform: rotate(0.5deg) translateZ(46px);\n    opacity: 0.69;\n  }\n"]);

  _templateObject2$5 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$9() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  width: 100%;\n  height: 100%;\n  position: absolute;\n  top: 0;\n  left: 0;\n  z-index: 201;\n  background: white;\n\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  flex-direction: column;\n  font-family: \"Lato\", sans-serif;\n  perspective: 500px;\n\n  transition: opacity 1.5s ease;\n  transition-delay: 2.25s;\n  ", "\n  ", "\n"]);

  _templateObject$9 = function _templateObject() {
    return data;
  };

  return data;
}
var Container$6 = styled__default['default'].div(_templateObject$9(), function (props) {
  return props.finished && "opacity: 0;";
}, function (props) {
  return props.finished && "pointer-events: none;";
});
var flickering = core.keyframes(_templateObject2$5());
var TextContainer = styled__default['default'].div(_templateObject3$4(), function (props) {
  return props.finished && "opacity: 0;";
});
var Text$2 = styled__default['default'].p(_templateObject4$2(), flickering);

var Loading = function Loading(props) {
  var progress = props.progress;
  return /*#__PURE__*/React.createElement(Container$6, {
    finished: progress === 100
  }, /*#__PURE__*/React.createElement(TextContainer, {
    finished: progress === 1
  }, /*#__PURE__*/React.createElement(Text$2, null, Math.floor(progress))), /*#__PURE__*/React.createElement(SpinningLoading, {
    progress: progress
  }));
};

var PortalLoadingScreen = function PortalLoadingScreen(props) {
  var setFixedPath = props.setFixedPath;

  var _useEnvironment = useEnvironment(),
      setPaused = _useEnvironment.setPaused;

  var _useState = react.useState(false),
      start = _useState[0],
      setStart = _useState[1];

  var beginExperience = function beginExperience() {
    setPaused(false);
    setStart(true);
  };

  var selfExplore = function selfExplore() {
    setFixedPath(false);
    beginExperience();
  };

  var progress = useControlledProgress();
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Loading, {
    progress: progress
  }), /*#__PURE__*/React.createElement(Transition, {
    start: start,
    beginExperience: beginExperience,
    selfExplore: selfExplore
  }));
};

var fetchAccount = /*#__PURE__*/function () {
  var _ref = _asyncToGenerator__default['default']( /*#__PURE__*/regeneratorRuntime.mark(function _callee(username) {
    var data, response, result;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.prev = 0;
            _context.next = 3;
            return fetch__default['default']("https://instagram.com/" + username + "/?__a=1");

          case 3:
            response = _context.sent;
            _context.next = 6;
            return response.json();

          case 6:
            result = _context.sent;
            data = result.graphql.user;
            _context.next = 13;
            break;

          case 10:
            _context.prev = 10;
            _context.t0 = _context["catch"](0);
            return _context.abrupt("return", _context.t0);

          case 13:
            return _context.abrupt("return", data);

          case 14:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[0, 10]]);
  }));

  return function fetchAccount(_x) {
    return _ref.apply(this, arguments);
  };
}();

/**
 * Hook to fetch portal from backend gien an ID.
 *
 * If it finds an instagram username it will make network call to fetch instagram
 *
 * Will build assets from network fetch into a modified version of object
 * using the passed in builder function (optional)
 *
 * @param id
 * @param portalBuilder
 */
var usePortal = function usePortal(id, portalBuilder) {
  var _useState = react.useState(),
      result = _useState[0],
      setResult = _useState[1];

  var _useState2 = react.useState(),
      error = _useState2[0],
      setError = _useState2[1];

  react.useEffect(function () {
    var fetchData = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator__default['default']( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var localResult, response, _result, _err$data, instagramUsername, instagramResponse;

        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                localResult = {}; // fetch portal

                _context.prev = 1;
                _context.next = 4;
                return fetch__default['default']("https://27m2mbw4h0.execute-api.us-west-1.amazonaws.com/fetch", {
                  method: "POST ",
                  body: JSON.stringify({
                    id: id
                  }),
                  headers: {
                    "Content-Type": "application/json"
                  }
                });

              case 4:
                response = _context.sent;
                _context.next = 7;
                return response.json();

              case 7:
                _result = _context.sent;
                Object.assign(localResult, _result);
                _context.next = 15;
                break;

              case 11:
                _context.prev = 11;
                _context.t0 = _context["catch"](1);
                setError(((_err$data = _context.t0.data) == null ? void 0 : _err$data.message) || "Could not find portal. Please contact us if the problem persists");
                return _context.abrupt("return");

              case 15:
                if (localResult) {
                  _context.next = 18;
                  break;
                }

                setError("Could not find portal. Please contact us if the problem persists");
                return _context.abrupt("return");

              case 18:
                // fetch instagram
                instagramUsername = localResult.instagramUsername;

                if (!instagramUsername) {
                  _context.next = 30;
                  break;
                }

                _context.prev = 20;
                _context.next = 23;
                return fetchAccount(instagramUsername);

              case 23:
                instagramResponse = _context.sent;
                localResult.instagram = instagramResponse;
                _context.next = 30;
                break;

              case 27:
                _context.prev = 27;
                _context.t1 = _context["catch"](20);
                console.error("Failed to fetch @" + instagramUsername);

              case 30:
                setResult(localResult);

              case 31:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, null, [[1, 11], [20, 27]]);
      }));

      return function fetchData() {
        return _ref.apply(this, arguments);
      };
    }();

    if (!result && !error) {
      fetchData();
    }
  }, [result, error]);
  var builtResult = portalBuilder && result ? portalBuilder(result) : result;
  return {
    result: builtResult,
    error: error
  };
};

var SHOW_PLAYER_HITBOX$1 = false;

/**
 * Player represents a physics-enabled player in the environment, complete with a
 * control scheme and a physical representation that interacts with other physics-
 * enabled objects.
 *
 * There should only be one player per environment.
 *
 * @constructor
 */
var TrackPlayer = function TrackPlayer(props) {
  var _props$initPos = props.initPos,
      initPos = _props$initPos === void 0 ? new THREE.Vector3(0, 1, 0) : _props$initPos,
      _props$initRot = props.initRot,
      initRot = _props$initRot === void 0 ? 0 : _props$initRot;

  var _useThree = reactThreeFiber.useThree(),
      camera = _useThree.camera;

  var _useEnvironment = useEnvironment(),
      setPlayer = _useEnvironment.setPlayer; // physical body


  var _useSphere = cannon.useSphere(function () {
    return {
      mass: 500,
      position: initPos.toArray(),
      args: 1,
      fixedRotation: true
    };
  }),
      bodyRef = _useSphere[0],
      bodyApi = _useSphere[1]; // producer


  var position = react.useRef(new THREE.Vector3(0, 0, 0));
  var velocity = react.useRef(new THREE.Vector3(0, 0, 0));
  var lockControls = react.useRef(false);
  var raycaster = react.useRef(new THREE.Raycaster(new THREE.Vector3(), new THREE.Vector3(), 0, 3)); // consumer

  var quaternion = react.useRef(new THREE.Quaternion(0, 0, 0, 0)); // rad on y axis

  var controls = react.useRef(); // setup player

  react.useEffect(function () {
    // store position and velocity
    bodyApi.position.subscribe(function (p) {
      position.current.set(p[0], p[1], p[2]);
    });
    bodyApi.velocity.subscribe(function (v) {
      return velocity.current.set(v[0], v[1], v[2]);
    });
    var xLook = initPos.x + 100 * Math.cos(initRot);
    var zLook = initPos.z + 100 * Math.sin(initRot);
    camera == null ? void 0 : camera.lookAt(xLook, initPos.y, zLook);

    if (reactDeviceDetect.isMobile) {
      window.addEventListener("click", function () {
        controls.current = new DeviceOrientationControls.DeviceOrientationControls(camera);
      });
    }

    setPlayer(createPlayerRef(bodyApi, position, velocity, lockControls, raycaster));
  }, []); // update player every frame

  reactThreeFiber.useFrame(function (_ref) {
    var clock = _ref.clock;

    if (reactDeviceDetect.isMobile && controls.current) {
      controls.current.update();
    }

    var dist = (22 + 50) / 2;
    var x = dist * Math.cos(clock.getElapsedTime() / 10);
    var z = dist * Math.sin(clock.getElapsedTime() / 10);
    camera.position.set(x, 2, z);
    return;
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, !reactDeviceDetect.isMobile && /*#__PURE__*/React.createElement(MouseFPSCamera, {
    quaternion: quaternion,
    position: position
  }), /*#__PURE__*/React.createElement("mesh", {
    ref: bodyRef,
    name: "player"
  }, SHOW_PLAYER_HITBOX$1 ));
};

function _templateObject2$6() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: absolute;\n  text-align: center;\n  color: red;\n  width: 100%;\n  max-width: 500px;\n  padding: 0 10%;\n  top: 50%;\n  left: 50%;\n  transform: translate(-50%, -50%);\n"]);

  _templateObject2$6 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$a() {
  var data = _taggedTemplateLiteralLoose__default['default'](["\n  position: absolute;\n  width: 100%;\n  height: 100%;\n  overflow: hidden;\n\n  canvas {\n    position: absolute;\n    top: 0;\n    right: 0;\n    bottom: 0;\n    left: 0;\n    outline: 0;\n  }\n"]);

  _templateObject$a = function _templateObject() {
    return data;
  };

  return data;
}
var Container$7 = styled__default['default'].div(_templateObject$a());
var ErrorText = styled__default['default'].p(_templateObject2$6());
var defaultCanvasProps$1 = {
  shadowMap: true,
  gl: {
    alpha: false
  },
  camera: {
    position: [0, 2, 0],
    near: 0.01,
    far: 200
  }
};
var defaultPhysicsProps$1 = {
  iterations: 20,
  size: 10,
  allowSleep: false,
  defaultContactMaterial: {
    friction: 0
  }
};

/**
 *
 * Provides an environment that loads a portal based on the url
 *
 * Player Type: First Person on a track or with WASD/Joystick
 * Physics: Enabled with default y=0 floor plane
 * Loading Screen: Spaces Portal Edition
 * Pause Menu: Spaces Edition
 *
 */
var PortalEnvironment = function PortalEnvironment(props) {
  var children = props.children,
      canvasProps = props.canvasProps,
      physicsProps = props.physicsProps,
      children2d = props.children2d,
      portalId = props.portalId,
      portalHandler = props.portalHandler;

  var _usePortal = usePortal(portalId, portalHandler),
      result = _usePortal.result,
      error = _usePortal.error;

  var state = useEnvironmentState();

  var localState = _extends__default['default']({}, state, {
    portal: result
  });

  var _useState = react.useState(true),
      fixedPath = _useState[0],
      setFixedPath = _useState[1];

  if (error) {
    return /*#__PURE__*/React.createElement(ErrorText, null, error);
  }

  return /*#__PURE__*/React.createElement(BrowserChecker, null, /*#__PURE__*/React.createElement(GlobalStyles, null), /*#__PURE__*/React.createElement(Container$7, {
    ref: state.containerRef
  }, /*#__PURE__*/React.createElement(reactThreeFiber.Canvas, _extends__default['default']({}, defaultCanvasProps$1, canvasProps), /*#__PURE__*/React.createElement(cannon.Physics, _extends__default['default']({}, defaultPhysicsProps$1, physicsProps), /*#__PURE__*/React.createElement(environmentStateContext.Provider, {
    value: localState
  }, fixedPath ? /*#__PURE__*/React.createElement(TrackPlayer, null) : /*#__PURE__*/React.createElement(TrackPlayer, {
    initPos: new THREE.Vector3(0, 2, 53)
  }), children))), /*#__PURE__*/React.createElement(environmentStateContext.Provider, {
    value: localState
  }, children2d, /*#__PURE__*/React.createElement(PortalLoadingScreen, {
    setFixedPath: setFixedPath
  }), /*#__PURE__*/React.createElement(DesktopPause, null), reactDeviceDetect.isMobile && /*#__PURE__*/React.createElement(MobilePause, null))));
};

exports.Arrow = Arrow;
exports.Audio = Audio;
exports.Background = Background;
exports.Floating = Floating;
exports.HDRI = HDRI;
exports.Image = Image;
exports.InfinitePlane = InfinitePlane;
exports.Interactable = Interactable;
exports.Logo = Logo;
exports.PortalEnvironment = PortalEnvironment;
exports.RealisticEffects = RealisticEffects;
exports.Shop = Shop;
exports.StandardEnvironment = StandardEnvironment;
exports.Text = Text;
exports.Video = Video;
exports.environmentStateContext = environmentStateContext;
exports.useControlledProgress = useControlledProgress;
exports.useConvexCollision = useConvexCollision;
exports.useEnvironment = useEnvironment;
exports.useEnvironmentState = useEnvironmentState;
exports.useShopify = useShopify;
exports.useTrimeshCollision = useTrimeshCollision;
import React from "react";
